File: Maximum Product Subarray.py
class Solution:

	# Function to find maximum
	# product subarray
	def maxProduct(self,arr):
		# code here
        max_product = arr[0]
        min_product = arr[0]
        result = arr[0]
        
        # Traverse the array starting from the second element
        for i in range(1, len(arr)):
            num = arr[i]
            
            # If the current number is negative, swap max and min
            if num < 0:
                max_product, min_product = min_product, max_product
            
            # Update max_product and min_product
            max_product = max(num, max_product * num)
            min_product = min(num, min_product * num)
            
            # Update the result
            result = max(result, max_product)
        
        return result

File: Maximum Product Subarray.cpp
class Solution {
  public:
    int maxProduct(vector<int> &arr) {
        int n=arr.size(),maxi=arr[0],mini=arr[0],ans=arr[0];
        
        for(int i=1;i<n;i++){
            if(arr[i]<0) swap(maxi,mini);
            maxi=max(arr[i],maxi*arr[i]);
            mini=min(arr[i],mini*arr[i]);
            
            ans=max(ans,maxi);
        }
        
        return ans;
    }
};
